### PR TITLE
Fix useZoomOut inserter behavior

### LINKED ISSUE #66328: Zoom out returns after inserter is closed
### Description

Disabling zoom out while the inserter is open makes it come back when the inserter is closed.

### Step-by-step reproduction instructions

https://github.com/user-attachments/assets/0efcb333-6eab-40e8-92e4-91c436535660

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

- [x] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [x] Yes

### Please confirm which theme type you used for testing.

- [x] Block
- [ ] Classic
- [ ] Hybrid (e.g. classic with theme.json)
- [ ] Not sure

### PR DESCRIPTION
**Pull request by jeryj**
Fixes https://github.com/WordPress/gutenberg/issues/66328

## What?

Track if the useZoomOut hook changed the zoom level for the user. If we never changed it, don't change the zoom level on unmount.

Request from @richtabor: If we did change zoom level via the `useZoomOut` hook (controlled mode) _always `resetZoomLevel()` on unmount_.

## Why?

Fixes a bug where zoom out was engaged/disengaged incorrectly on unmount. 

## How?

Adds a ` controlZoomLevelRef`  that tracks if we ever changed the zoom level for the user. When unmounting, if we changed the zoom level, reset it _unless they manually toggled the zoom level after us_. In that condition, `controlZoomLevelRef` is set to `false`.

## Testing Instructions

The key concepts here are:
- If we most recently changed the zoom level for them (in or out), we always resetZoomLevel() level when unmounting.
- If the user most recently changed the zoom level (manually toggling), we do nothing when unmounting.
 
So, with these in mind, we have two scenarios to watch for:
- Closing the inserter always resets the zoom level if the inserter changed the zoom level (clicking on blocks tab from zoomed out state OR clicking on patterns or media tab from zoomed in state)
- Closing the inserter NEVER affects zoom level if zoom state was manually toggled via the zoom out button. However, if you manually toggle, then click to a tab that sets the zoom state, we are now controlling it again.

1. Enter Zoom out manually
2. Open the inserter.
3. The pattern tab should be selected.
4. Exit zoom out manually.
5. Close the inserter.
6. Editor should remain in non-zoomed state (regular editing)

From full screen mode
1. Open inserter
2. Click patterns tab
3. Zoom out should engage
4. Close inserter
5. Zoom out should disengage

From full screen mode, exit zoom out manually
1. Open inserter
2. Click patterns tab
3. Zoom out should engage
4. Manually exit zoom out
5. Close inserter
6. Zoom out should not engage (remain in full screen mode)

## Screen Recording

https://github.com/user-attachments/assets/082ce511-c038-4de4-aa92-5c35f7ef6a5f

### CODE REVIEW

File: packages/block-editor/src/hooks/use-zoom-out.js [hunk 1]
```diff
@@ -12,32 +12,53 @@ import { unlock } from '../lock-unlock';
...
+ * @param {boolean} enabled If we should enter into zoomOut mode or not
  */
-export function useZoomOut( zoomOut = true ) {
+export function useZoomOut( enabled = true ) {
 	const { setZoomLevel, resetZoomLevel } = unlock(
 		useDispatch( blockEditorStore )
 	);
 	const { isZoomOut } = unlock( useSelect( blockEditorStore ) );
 
+	const isZoomedOut = isZoomOut();
```
Comment (Mamaduka) [inline packages/block-editor/src/hooks/use-zoom-out.js]: Calling a selector during the render isn't recommended.

Comment (jeryj) [reply packages/block-editor/src/hooks/use-zoom-out.js]: Thanks for catching! Updated to call it within a `useSelect`.

**Comment by richtabor** [comment]
Experience wise, this feels nice. It stays zoomed out when expected to, and zooms back in when appropriate. 

https://github.com/user-attachments/assets/a4ef232c-4b07-47e3-8b4e-cdf81565a567

**Review by richtabor — APPROVED** [review]
This feels nice! You don't end up in unexpected scenarios where zoom out is auto re-engaged.

File: test/e2e/specs/site-editor/site-editor-inserter.spec.js [hunk 2]
```diff
@@ -64,4 +62,396 @@ test.describe( 'Site Editor Inserter', () => {
...
+
+	test( 'should open the inserter to patterns tab if using zoom out and stay in zoom out when closing the inserter', async ( {
+		InserterUtils,
+	} ) => {
+		const zoomOutButton = InserterUtils.getZoomOutButton();
+		const inserterButton = InserterUtils.getInserterButton();
+		const blockLibrary = InserterUtils.getBlockLibrary();
+
+		await zoomOutButton.click();
+		await expect( await InserterUtils.getZoomCanvas() ).toBeVisible();
```
Comment (Mamaduka) [inline test/e2e/specs/site-editor/site-editor-inserter.spec.js]:
```suggestion
		await expect( InserterUtils.getZoomCanvas() ).toBeVisible();
```

I think there's no need to `await` canvas locator. Usually, you would write it like this - `await expect( page.locator( '.is-zoomed-out' ) ).toBeVisible();`.

Linter will remove extra await if you try to put it there.

File: test/e2e/specs/site-editor/site-editor-inserter.spec.js [hunk 3]
```diff
@@ -64,4 +62,396 @@ test.describe( 'Site Editor Inserter', () => {
...
+		await inserterButton.click();
+
+		await expect( blockLibrary ).toBeHidden();
+
+		// We should not return to zoom in since it was manually disengaged
+		await expect( await InserterUtils.getZoomCanvas() ).toBeVisible();
+	} );
 } );
+
+class InserterUtils {
```
Comment (Mamaduka) [inline test/e2e/specs/site-editor/site-editor-inserter.spec.js:259]: IMO, the utils class here seems redundant. It just holds aliases for simple locator calls. Why not just inline them?

**Review by Mamaduka — APPROVED** [review]
Thank you, @jeryj!

The fix looks good to me ✅ I just left a couple of notes regarding e2e tests.

P.S. Rebasing should fix the failing e2e test.

File: test/e2e/specs/site-editor/site-editor-inserter.spec.js (hunk 3 above)
Comment (jeryj) [reply test/e2e/specs/site-editor/site-editor-inserter.spec.js:259]: I redid them today to make them a lot more readable. They were pretty scattered and hard to piece together IMO since there are so many conditions to test.

I'm still using the `InserterUtils` and also added a `ZoomUtils`. They do hold basic locators and some `expects`, but I think it make it way more readable. I was able to remove a lot of comments and just have the method names document the actions.

Comment (Mamaduka) [reply test/e2e/specs/site-editor/site-editor-inserter.spec.js:259]: Makes sense!

**Comment by jeryj** [comment]
The same e2e test keeps failing even after rebasing and trying several times to rerun it. It passes locally. Not sure what's going on with it. It doesn't seem related to this PR.

File: test/e2e/specs/site-editor/preload.spec.js [hunk 4]
```diff
@@ -6,6 +6,7 @@ const { test, expect } = require( '@wordpress/e2e-test-utils-playwright' );
 test.describe( 'Preload', () => {
 	test.beforeAll( async ( { requestUtils } ) => {
 		await requestUtils.activateTheme( 'emptytheme' );
+		await requestUtils.resetPreferences();
```
Comment (jeryj) [inline test/e2e/specs/site-editor/preload.spec.js:9]: This should really be in a different PR, but I wanted to see if adding this would fix the e2e issue I'm running into. [More context in the PR that implemented the Preload test](https://github.com/WordPress/gutenberg/pull/67497#issuecomment-2523847534).

tl;dr: The Preload test will have different results depending on what runs before it, and this should help normalize it. I just happened to run into the issue by having my test run on the same shard.

Comment (jeryj) [reply test/e2e/specs/site-editor/preload.spec.js:9]: It did work -- I would have liked this fix a different PR but it auto-merged once the tests passed.
